### ISSUE TITLE
What does global orient mean?

### ISSUE THREAD

**Issue by hahamini** [geopavlakos/hamer#66]
I'm asking this question because I want to calculate the vector of the palm from the wrist(root).

I guess it's the orientation of the root keypoint. Is that right?
If so, what is the reference of that coordinate system? (For example, palm orientation is z-axis)

Thank you.

**Comment by geopavlakos** [comment]
The parameter `global_orient` corresponds to the global orientation of the root of the MANO model.
About the reference, you will need to check the representation of MANO internally.

**Comment by hahamini** [comment]
@geopavlakos 
Thanks for your answer.

I have a few more questions.
1. I'm using hand_pose and pred_keypoints_3d to project a pose to a 2D image.
Here, the right hand is projected well, but the left hand is projected to the wrong location. I don't know why.
2. What is the index correspondence between hand_pose, a rotation matrix with 15 members, and keypoints_3d, a matrix with 21 members?
3. Is the pose of each joint relative to its parent node?
For example, if root, 1(thumb pip), 2(thumb dip), 3(thumb tip), I wonder if R_root_to_tip = R1@R2@R3 or R3.

Here is my demo code.
```python
           for n in range(batch_size):
                global_orient = (
                    out["pred_mano_params"]["global_orient"].detach().cpu().numpy()[n]
                )
                hand_pose = (
                    out["pred_mano_params"]["hand_pose"].detach().cpu().numpy()[n]
                )
                keypoints_3d = out["pred_keypoints_3d"][n].detach().cpu().numpy() + (
                    pred_cam_t_full[n]
                )
                # Hand root
                cv2.drawFrameAxes(
                    canvas,
                    intrinsic.camera_matrix,
                    np.zeros(5),
                    cv2.Rodrigues(np.squeeze(global_orient))[0],
                    keypoints_3d[0],
                    0.01,
                )
                # Thumb pip
                cv2.drawFrameAxes(
                    canvas,
                    intrinsic.camera_matrix,
                    np.zeros(5),
                    cv2.Rodrigues(hand_pose[0])[0],
                    keypoints_3d[2],
                    0.01,
                )
                # Thumb tip
                cv2.drawFrameAxes(
                    canvas,
                    intrinsic.camera_matrix,
                    np.zeros(5),
                    cv2.Rodrigues(hand_pose[2])[0],
                    keypoints_3d[4],
                    0.01,
                )
```

**Comment by hahamini** [comment]
@geopavlakos This was really helpful, thank you.

**Comment by geopavlakos** [comment]
You should multiply the first dimension with -1, not the second and the third.
Similarly, you need to multiply the first dimension of `pred_cam_t_full` with -1 before you add them together.

**Comment by hahamini** [comment]
@geopavlakos 
As you said, I multiplied the first dimension (pred_keypoints_3d[:,0]) of pred_keypoints_3d(21,3) by -1, and also multiplied the first dimension (pred_cam_t_full[0]) of pred_cam_t_full(3,) by -1. 
However, when I verified it with re-projection, the result was wrong.
(Of course, the right hand result is perfect.)

```python
                global_orient = (
                    out["pred_mano_params"]["global_orient"].detach().cpu().numpy()[n]
                )
                hand_pose = (
                    out["pred_mano_params"]["hand_pose"].detach().cpu().numpy()[n]
                )
                keypoints_3d = out["pred_keypoints_3d"].detach().cpu().numpy()[n]
                # keypoints_3d -> (21,3)
                # pred_cam_t_full[n] -> (3,)
                if right[n] == 0:
                    keypoints_3d[:, 0] *= -1
                    pred_cam_t_full[n][0] *= -1
                    # flip_matrix = np.array([[1, 0, 0], [0, -1, 0], [0, 0, -1]])
                    # keypoints_3d = np.matmul(
                    #     flip_matrix, keypoints_3d.reshape(-1, 3, 1)
                    # ).reshape(-1, 3)
                keypoints_3d += pred_cam_t_full[n]
```

**Comment by geopavlakos** [comment]
My bad, the multiplication for the camera translation has already been taken care of in line 138 of the demo. You only need to multiply the first dimension of the keypoints with -1 and that should be it.

**Comment by hahamini** [comment]
@geopavlakos 
Does that mean I don't have to change anything in my code to handle the right hand?

I just need to add pred_cam_t_full to get the 3D keypoints for the right hand?
But still the result was wrong.

```
                global_orient = (
                    out["pred_mano_params"]["global_orient"].detach().cpu().numpy()[n]
                )
                hand_pose = (
                    out["pred_mano_params"]["hand_pose"].detach().cpu().numpy()[n]
                )
                keypoints_3d = out["pred_keypoints_3d"].detach().cpu().numpy()[n]
                # keypoints_3d -> (21,3)
                # pred_cam_t_full[n] -> (3,)
                #if right[n] == 0:
                #    keypoints_3d[:, 0] *= -1
                #    pred_cam_t_full[n][0] *= -1
                keypoints_3d += pred_cam_t_full[n]
```

**Comment by geopavlakos** [comment]
I mentioned above that you "only need to multiply the first dimension of the keypoints with -1". By commenting out the whole if-statement in your code, you don't multiply the keypoint of the left hand with -1. The correct code would be:
```
                if right[n] == 0:
                    keypoints_3d[:, 0] *= -1
                keypoints_3d += pred_cam_t_full[n]
```
